### PR TITLE
android: add listener for selection change

### PR DESCRIPTION
**Pull request by grosalex**
Add a callback on selection change from the webview to trigger the context menu in the app

<details>
<summary>REVIEWERS (4)</summary>

R=@asaudemont
R=@christophespy007
R=@MickaelGuilbeaud
R=@prcaen
</details>

Manual testing: 🚫

### CODE REVIEW

**Review by MickaelGuilbeaud — APPROVED** [review]
LGTM. There's a few dev logs left in this PR.

File: library/src/main/assets/selection.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+var common = {
```
Comment (asaudemont) [inline library/src/main/assets/selection.js]: We should drop `common` (which doesn't really make sense here) and put everything into a `selection` object:

```js
var selection = {
    onSelectionChange: function() {
        // ...
    }

    notifySelection: function() {
        // ...
    }

    setup: function() {
        // ...
    }
};
```

and in `SourceUtils.getScriptPageHeader`:

```java
return ... +
    "    <script src=\"./selection.js\"></script>\n" +
    "    <script>selection.setup();</script>\n" +
    ...
```

?

File: library/src/main/java/com/pddstudio/highlightjs/models/SelectionCallback.kt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.pddstudio.highlightjs.models
+
+interface SelectionCallback {
+    fun onSelectionChange(selection: String?)
```
Comment (asaudemont) [inline library/src/main/java/com/pddstudio/highlightjs/models/SelectionCallback.kt]: Rename `selection` to `selectedText`?

File: library/src/main/java/com/pddstudio/highlightjs/utils/SourceUtils.java [hunk 3]
```diff
@@ -44,6 +44,8 @@ private static String getScriptPageHeader(boolean showLineNumbers) {
 		return "    <script src=\"./highlight.pack.js\"></script>\n" +
 				(showLineNumbers ? "<script src=\"./highlightjs-line-numbers.min.js\"></script>\n" : "") +
 				"    <script>hljs.initHighlightingOnLoad();</script>\n" +
+				"<script src=\"./selection.js\"></script>\n" +
```
Comment (asaudemont) [inline library/src/main/java/com/pddstudio/highlightjs/utils/SourceUtils.java]:
```suggestion
				"    <script src=\"./selection.js\"></script>\n" +
```

File: library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt [hunk 4]
```diff
@@ -256,4 +263,16 @@ class HighlightJsView : WebView, FileUtils.Callback {
...
+
+    inner class JsInterface() {
+        @JavascriptInterface
+        fun onSelectionChange(value: String?) {
+            val start = value?.let { content?.indexOf(it) }
+            onSelectionChange?.onSelectionChange(
+                start?.let {
+                    content?.subSequence(it, it + value.length).toString()
+                }
+            )
```
Comment (asaudemont) [inline library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt]:
```suggestion
            onSelectionChange?.onSelectionChange(value)
```

?

File: library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt [hunk 5]
```diff
@@ -39,6 +42,8 @@ class HighlightJsView : WebView, FileUtils.Callback {
     private var onLanguageChangedListener: OnLanguageChangedListener? = null
     private var onThemeChangedListener: OnThemeChangedListener? = null
     private var onContentChangedListener: OnContentChangedListener? = null
+    var onSelectionChange: SelectionCallback? = null
```
Comment (asaudemont) [inline library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt]: Either:

- Rename `HighlightJsView.onSelectionChange` to `selectionCallback` (for consistency with the `SelectionCallback` type).

or:

- Change the type of `onSelectionChange` to `((String?) -> Unit)?` and delete `SelectionCallback`.

?

(My preference goes to making `onSelectionChange` a function type. And also to renaming it to `onSelectionChange*d*`.)

File: library/src/main/assets/selection.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            var selection = document.getSelection();
+
+            if (selection.anchorNode == null || selection.focusNode == null || selection.isCollapsed) {
+                jsBridge.onSelectionChange(null);
+                return;
+            }
+            jsBridge.onSelectionChange(
+                selection.toString()
+            );
+        },
```
Comment (asaudemont) [inline library/src/main/assets/selection.js]: Fix indentation?

File: library/src/main/assets/selection.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    notifySelection: function() {
+            var selection = document.getSelection();
+
+            if (selection.anchorNode == null || selection.focusNode == null || selection.isCollapsed) {
+                jsBridge.onSelectionChange(null);
+                return;
+            }
+            jsBridge.onSelectionChange(
+                selection.toString()
+            );
```
Comment (asaudemont) [inline library/src/main/assets/selection.js]: nit:

```js
if (selection.anchorNode == null || selection.focusNode == null || selection.isCollapsed) {
    jsBridge.onSelectionChange(null);
} else {
    jsBridge.onSelectionChange(selection.toString());
}
```

?

File: library/src/main/java/com/pddstudio/highlightjs/utils/SourceUtils.java [hunk 8]
```diff
@@ -44,6 +44,8 @@ private static String getScriptPageHeader(boolean showLineNumbers) {
 		return "    <script src=\"./highlight.pack.js\"></script>\n" +
 				(showLineNumbers ? "<script src=\"./highlightjs-line-numbers.min.js\"></script>\n" : "") +
 				"    <script>hljs.initHighlightingOnLoad();</script>\n" +
+				"	 <script src=\"./selection.js\"></script>\n" +
+				"    <script>selection.setup();</script>\n" +
```
Comment (asaudemont) [inline library/src/main/java/com/pddstudio/highlightjs/utils/SourceUtils.java:48]:
```suggestion
				"    <script src=\"./selection.js\"></script>\n" +
				"    <script>selection.setup();</script>\n" +
```

File: library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt [hunk 9]
```diff
@@ -39,6 +37,8 @@ class HighlightJsView : WebView, FileUtils.Callback {
     private var onLanguageChangedListener: OnLanguageChangedListener? = null
     private var onThemeChangedListener: OnThemeChangedListener? = null
     private var onContentChangedListener: OnContentChangedListener? = null
+    var selectionCallback: SelectionCallback? = null
```
Comment (asaudemont) [inline library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt:41]:
```suggestion

    var selectionCallback: SelectionCallback? = null
```

?

File: library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt [hunk 10]
```diff
@@ -256,4 +258,13 @@ class HighlightJsView : WebView, FileUtils.Callback {
     fun setShowLineNumbers(showLineNumbers: Boolean) {
         this.showLineNumbers = showLineNumbers
     }
+
+    inner class JsInterface() {
+        @JavascriptInterface
+        fun onSelectionChange(value: String?) {
+            selectionCallback?.onSelectionChange(
+                value
+            )
```
Comment (asaudemont) [inline library/src/main/java/com/pddstudio/highlightjs/HighlightJsView.kt]:
```suggestion
            selectionCallback?.onSelectionChange(value)
```

?

**Review by asaudemont — APPROVED** [review]

**Pull request merged**